### PR TITLE
Data Mining Fixes

### LINKED ISSUE #265: Data Mining Post Process Does Not Copy Input Set to Output
--------
Issue Description
--------
##### What did you expect to see happen?
When running a kMeans clustering model in a PostProcess step, I specified two separate PointSets (crescents_data and cluster_results) for input and output. I then plotted cluster_results using a cluster plotter in an IOStep. I expected a plot of the clusters in my data.

##### What did you see instead?
The application returned that my input space for the plotter (cluster_results|Input) was empty and wouldn't plot anything. To work around this, I had to specify that my input and output for the PostProcess step were both crescents_data and then fed this to the plotter. It then plotted correctly.

##### Do you have a suggested fix for the development team?
The PostProcessor should return a full set of data to the output set.

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

[data.zip](https://github.com/idaholab/raven/files/1163110/data.zip)

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #289: DataMining Plots
--------
Issue Description
--------

There is an issue in the `<type>dataMining</type>` plot available in RAVEN. It appears that if the data is written to file, it exists, however when plotting the data, it will skip data points with a -1 label (signifies an outlier in sklearn). These points should be rendered differently to ensure that the user does not mistake them for a single cluster. Perhaps by modifying the marker/glyph in some way.

My first thought would be to make them "hollowed out" points, but it occurs to me that they would just look white. If they are sized smaller or use a different marker, the hope is that they will be different enough to not register in a user's head as part of a single cluster.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by maljovec**
Closes #289 and Closes #265.

* Adding the capability to copy information from the input to the output data object for a DataMining PostProcessor

* Fixing the data mining plot to show outliers (label=-1).

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

**Comment by maljovec** [comment]
@mandd, I may need some help in figuring out how to handle the HistorySet's for this feature.

File: framework/OutStreams/OutStreamPlot.py [hunk 1]
```diff
@@ -1772,12 +1772,54 @@ def addOutput(self):
                   for k, col in zip(range(int(clusterDict[pltindex]['noClusters'])), colors):
                     myMembers = self.clusterValues[pltindex][1][0] == k
                     self.actPlot = plt.scatter(clusterDict[pltindex]['clusterValues'][myMembers, 0], clusterDict[pltindex]['clusterValues'][myMembers, 1] , color = col, **dataMiningPlotOptions)
+
+                  ## Handle all of the outlying data
+                  myMembers = self.clusterValues[pltindex][1][0] == -1
+                  ## resize the points
+                  dataMiningPlotOptions['s'] /= 2
```
Comment (mandd) [inline framework/OutStreams/OutStreamPlot.py]: I think I don't have clear the /= and *= trick here :-(

File: framework/PostProcessors/DataMining.py [hunk 2]
```diff
@@ -347,40 +347,118 @@ def _localReadMoreXML(self, xmlNode):
...
+          if key in inputObject.getParaKeys('Input'):
+            column = {}
+            for historyIdx in inputObject.getParametersValues('inputs').keys():
+              column[historyIdx] = copy.copy(inputObject.getParam('Input', historyIdx)[key])
+            # column = copy.copy(inputObject.getParam('Input', key))
+          elif key in inputObject.getParaKeys('Output'):
+            column = {}
+            for historyIdx in inputObject.getParametersValues('outputs').keys():
+              column[historyIdx] = copy.copy(inputObject.getParam('Output', historyIdx)[key])
+            # column = copy.copy(inputObject.getParam('Output', key))
```
Comment (mandd) [inline framework/PostProcessors/DataMining.py]: should we check if the user swithces a variable from input to output?

File: framework/OutStreams/OutStreamPlot.py (hunk 1 above)
Comment (maljovec) [inline framework/OutStreams/OutStreamPlot.py]: 's' is coming from matplotlib's [scatter](https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.scatter.html#matplotlib.axes.Axes.scatter). It represents the size of the marker. Here I am reducing the size of the marker to render the "outliers" (the points with labels < 0 as determined by the sklearn algorithms).

File: framework/OutStreams/OutStreamPlot.py [hunk 3]
```diff
@@ -1772,12 +1772,54 @@ def addOutput(self):
...
+                  ## and hollow out their markers
+                  if 'facecolors' in dataMiningPlotOptions:
+                    faceColors = dataMiningPlotOptions['facecolors']
+                  else:
+                    faceColors = None
+                  dataMiningPlotOptions['facecolors'] = 'none'
+
+                  self.actPlot = plt.scatter(clusterDict[pltindex]['clusterValues'][myMembers, 0], clusterDict[pltindex]['clusterValues'][myMembers, 1] , color = '#000000', **dataMiningPlotOptions)
+
+                  ## Restore the plot options to their original values
```
Comment (maljovec) [inline framework/OutStreams/OutStreamPlot.py]: This restores the point size to its initial value so any remaining plot calls use the user-specified or default value not the reduced size points.

File: framework/PostProcessors/DataMining.py (hunk 2 above)
Comment (maljovec) [inline framework/PostProcessors/DataMining.py]: Oops, copy and paste error here. Nice catch.

File: framework/PostProcessors/DataMining.py [hunk 4]
```diff
@@ -347,40 +347,118 @@ def _localReadMoreXML(self, xmlNode):
...
+
+          if column is None:
+            missingKeys.append(key)
+          else:
+            for historyIdx,value in column.items():
+              if key in outputObject.getParaKeys('Input'):
```
Comment (maljovec) [inline framework/PostProcessors/DataMining.py]: Here is where we determine where the extracted data from the input object goes into the output object.

**Comment by mandd** [comment]
Ready to merge

**Pull request merged**